src/metabase/lib/common.cljc

```diff
@@ -28,21 +32,31 @@
    (defmacro defop
      "Defines a clause creating function with given args.
       Calling the clause without query and stage produces a fn that can be resolved later."
-     [op-name argvec]
+     [op-name & argvecs]
      {:pre [(symbol? op-name)
-            (vector? argvec) (every? symbol? argvec)
-            (not-any? #{'query 'stage-number} argvec)]}
-     (let [arglist-expr (if (contains? (set argvec) '&)
-                          (cons 'list* (remove #{'&} argvec))
-                          argvec)]
-       `(mu/defn ~op-name :- [:or fn? ~(keyword "mbql.clause" (name op-name))]
-          ~(format "Create a clause of type `%s`." (name op-name))
-          ([~@argvec]
-           (fn [~'query ~'stage-number]
-             (~op-name ~'query ~'stage-number ~@argvec)))
-          ([~'query ~'stage-number ~@argvec]
-           (-> (into [~(keyword op-name)]
-                     (map (fn [~'arg]
-                            (->op-arg ~'query ~'stage-number ~'arg)))
-                     ~arglist-expr)
-               lib.options/ensure-uuid))))))
+            (every? vector? argvecs) (every? #(every? symbol? %) argvecs)
+            (every? #(not-any? #{'query 'stage-number} %) argvecs)]}
+     `(do
+          (mu/defn ~(symbol (str (name op-name) "-clause")) :- ~(keyword "mbql.clause" (name op-name))
+            ~(format "Create a standalone clause of type `%s`." (name op-name))
+            ~@(for [argvec argvecs
+                    :let [arglist-expr (if (contains? (set argvec) '&)
+                                         (cons `list* (remove #{'&} argvec))
+                                         argvec)]]
+                `([~'query ~'stage-number ~@argvec]
+                  (-> (into [~(keyword op-name)]
+                            (map (fn [~'arg]
+                                   (->op-arg ~'query ~'stage-number ~'arg)))
+                            ~arglist-expr)
+                      lib.options/ensure-uuid))))
+
+          (mu/defn ~op-name :- fn?
+            ~(format "Create a closure of clause of type `%s`." (name op-name))
+            ~@(for [argvec argvecs
+                    :let [arglist-expr (if (contains? (set argvec) '&)
+                                         (filterv (complement #{'&}) argvec)
+                                         (conj argvec []))]]
```

Comment:
With filters I made an attempt not to generate calls with `apply` if there were no varargs in play.